packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+import AsyncWriter from '@mongosh/async-rewriter';
+
+/**
+ * Anything to do with the internal shell state is stored here.
+ */
+export default class ShellInternalState {
+  public currentCursor: Cursor | AggregationCursor;
+  public currentDb: Database;
+  public messageBus: EventEmitter;
+  public asyncWriter: AsyncWriter;
+  public initialServiceProvider: ServiceProvider; // the initial service provider
+  public uri: string;
+  public connectionInfo: any;
+  constructor(initialServiceProvider: ServiceProvider, messageBus: any = new EventEmitter()) {
+    this.initialServiceProvider = initialServiceProvider;
+    this.messageBus = messageBus;
+    this.asyncWriter = new AsyncWriter(signatures);
+    const mongo = new Mongo(this);
+    this.currentCursor = null;
+    this.currentDb = mongo.getDB('test'); // TODO: set to CLI arg
+  }
+
+  async fetchConnectionInfo(): Promise<void> {
+    this.connectionInfo = await this.currentDb.mongo.serviceProvider.getConnectionInfo();
+    this.messageBus.emit('mongosh:connect', this.connectionInfo.extraInfo);
+  }
+
+  // TODO: do we need to close all Mongos ever made?
+  close(p): void {
+    try {
+      this.currentDb.mongo.serviceProvider.close(p);
+      // eslint-disable-next-line no-empty
+    } catch (e) {
+    }
+  }
+
+  use(db): any {
+    return this.currentDb.mongo.use(db);
+  }
+  show(arg): any {
+    return this.currentDb.mongo.show(arg);
+  }
+  async it(): Promise<any> {
+    if (!this.currentCursor) {
+      // TODO: warn here
+      return new CursorIterationResult();
+    }
+    return await this.currentCursor._it();
+  }
+  public help(): Help {
+    this.messageBus.emit('mongosh:help');
+    return new Help({
+      help: 'shell-api.help.description',
+      docs: 'https://docs.mongodb.com/manual/reference/method',
+      attr: [
+        {
+          name: 'use',
+          description: 'shell-api.help.help.use'
+        },
+        {
+          name: 'it',
+          description: 'shell-api.help.help.it'
+        },
+        {
+          name: 'show databases',
+          description: 'shell-api.help.help.show-databases'
+        },
+        {
+          name: 'show collections',
+          description: 'shell-api.help.help.show-collections'
+        },
+        {
+          name: '.exit',
+          description: 'shell-api.help.help.exit'
+        }
+      ]
+    });
+  }
+
+  /**
+   * Prepare a `contextObject` as global context and set it as context
+   * Add each attribute to the AsyncWriter also.
+   *
+   * The `contextObject` is prepared so that it can be used as global object
+   * for the repl evaluationi.
+   *
+   * @note The `contextObject` is mutated, it will retain all of its existing
+   * properties but also have the global shell api objects and functions.
+   *
+   * @param {Object} contextObject - contextObject an object used as global context.
+   */
+  setCtx(contextObject: any): void {
```

Comment:
I'd love to have a dedicated class `TopLevelContext` with decorators to keep
`use, show, it, db, rs, Mongo, printjson, DBQuery ... all bson types` etc ..., 
in the exact same way it happens for all the other methods and apis.

For example `Decimal128` is only supported since 3.4, wereas `Symbol` is deprecated
and would be probably unavailable in future versions, same could happen about other
objects that would not make sense for compatibility going further.

Another benefit of having these methods in a similar way we get for Collection etc
is that we could add those to the help automatically and mark them as commands 
if we want (with a new annotation @command).

```
class TopLevelContext { // or ShellGlobals
  // ... all the BSON classes ()

  get ObjectId {
    return ShellBSON.ObjectId(...args) // or just define this here
  }

  get DBQuery() {
    return Cursor;
  } 

  @command
  it()

  print() {}

  printjson() {
    return this.print()
  }

  @returnPromise
  connect() {

  }

  @serverVersion(['3.4', '...'])
  get Decimal128 {

  }
}
```

In `internalState` and `setCtx` we can just do this: 

```
constructor() {
  this.topLevelContext = new TopLevelContext(serviceProvider, mongo, 'etc ...');
}

setCtx() {
  Object.entries(signatures.TopLevelContext.attributes).forEach(name => {
    contextObject[name] = typeof this.topLevelContext[name] === 'function' ?
      this.topLevelContext[name].bind(this) : this.topLevelContext[name];
  });
}
```

In shell-evaluator instead we can replace the switch with:

```
if(signatures.TopLevelContext.commands.includes(cmd)) {
  return this.internalState[cmd](...argv);
}
```

Reply:
I don't understand what you mean by @command, can you explain or give an example?

Reply:
I moved `use`, `show`, and `it` into a top level object `ShellApi`. I got pretty far in moving the BSON types into that class as well (in this PR: https://github.com/mongodb-js/mongosh/tree/bson-pr) but realized it would affect the java-shell and I think it would be quite annoying to refactor that. So I'm going to say we should take this request and make it into a ticket to be worked on as part of the `dev prod` epic.